File: ra-idp/src/authentication/domain/authentication-context.ts
/**
 * Authentication component boundary.
 *
 * OAuth2/OIDC use cases consume this context and do not inspect password,
 * user lookup, or session-cookie details directly.
 */

export interface AuthenticationContext {
  sub: string
  auth_time: number
  amr: string[]
  acr?: string
  session_id?: string
  /** 追加 factor 検証待ちなら true。/authorize 完了経路は未認証扱いとし対応する factor challenge に誘導する。 */
  authentication_pending?: boolean
}

export interface AuthenticationContextResolver {
  resolve(headers: Headers): Promise<AuthenticationContext | null>
}

export class AuthenticationContextError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'AuthenticationContextError'
  }
}

File: ra-idp-go/internal/authentication/domain/authentication_context.go
// Authentication bounded context の境界。OAuth2/OIDC ユースケースはこの context を消費するだけで、
// password 検証・user lookup・session cookie の詳細には踏み込まない。
package domain

import "context"

type AuthenticationContext struct {
	Sub                   string
	AuthTime              int64
	AMR                   []string
	ACR                   string
	SessionID             string
	AuthenticationPending bool
	// StepUpAt は直近の step-up 再認証時刻 (Unix 秒、未実施は 0)。高 sensitivity 操作の
	// recency gate (ADR-043) が AuthTime と合わせて評価する。
	StepUpAt int64
}

type AuthenticationContextResolver interface {
	Resolve(ctx context.Context, headers Headers) (*AuthenticationContext, error)
}

// Headers は HTTP framework 非依存の薄い抽象 (key → first value)。
type Headers interface {
	Get(key string) string
}

// HTTPHeadersAdapter は標準 http.Header から Headers への変換。
type HTTPHeadersAdapter struct {
	H interface{ Get(string) string }
}

func (h HTTPHeadersAdapter) Get(k string) string { return h.H.Get(k) }
